README.md
# glmgen

Fork of the [Open Modeling Framework (OMF)](https://github.com/dpinney/omf) that has been pruned down to just its functionality for editing glm files, and then further developed in that direction.

#### Install

`pip install git+https://github.com/NREL/glmgen.git@master`

or 

`pip install git+https://github.com/NREL/glmgen.git@v4.1.0`

Previous versions are listed at https://github.com/NREL/glmgen/releases.

Unfortunately on Windows, while this command nominally works (after running it, `pip list` will show glmgen as installed), it [exits with errors](http://stackoverflow.com/q/23938896/1470262).

#### Starting points

- glmgen.feeder.GlmFile
- glmgen.makeGLM.makeGLM

This code is used in a research project that primarily uses the [PNNL Taxonomy Feeders](http://sourceforge.net/p/gridlab-d/code/HEAD/tree/Taxonomy_Feeders/) ([doc](http://www.gridlabd.org/models/feeders/taxonomy_of_prototypical_feeders.pdf)) as a starting point to develop feeder models populated with houses and solar.

#### Uninstall

`pip uninstall glmgen`



CHANGES.txt
v4.1.0, 08/26/2015 -- Previously populated feeder models can now have solar appended to them by setting 
                      glmgen.makeGLM.makeGLM's model_opts['append_solar'] parameter to True.
                      Refined street lighting schedule to be seasonal and to use a +/- 1 hour skew.
                      Made the solar inverter power factor a parameter (instead of being hard-coded to 1).
v4.0.0, 06/10/2015 -- Continued refactor of Configuration, and the residential, commercial, and 
                      PV population scripts. Buildings are added based on original load, and peak 
                      intensity per unit floor area. PV is now added based on the building floor 
                      area and annual energy intensity per unit floor area, with randomized azimuth 
                      and tilt, rather than a uniform 5 kW with default azimuth and tilt system.
v3.0.0, 05/11/2015 -- Refactored the Configuration module and its dependencies to improve the 
                      replacement of loads with house models. 
v2.3.0, 02/20/2014 -- GlmFile methods for setting the clock and voltage player files.
                      (Optional) overload of __setitem__ and __delitem__ to support 
                      inserts and dense keying.
                      Bug fix for not using homes while specifying one of the solar 
                      technologies.
v2.2.0, 01/16/2015 -- Added option to put .keep file in csv_output folder
v2.1.0, 12/31/2014 -- Added 'no voltage players' feature. Rescaled voltage player files.
v2.0.0, 12/01/2014 -- Final changes from FY14. Includes API changes in makeGLM
                      and Milsoft_GridLAB_D_Feeder_Generation.
v1.1.0, 07/23/2014 -- Added measured values to the slack bus to support co-simulation with transmission.
v1.0.1, 07/14/2014 -- Improved glm parsing speed by storing tokenList in a deque instead of a list.
v1.0.0, 07/07/2014 -- Started refactoring feeder.py to be more object-oriented. 
                      Broke the old write API (use GlmFile.__str__ and GlmFile.save instead). 
                      Continued to fix bugs and add externally-specified options. 
v0.1.0, 5-6/2014   -- Initial (informal) release.


glmgen/OMFmain.py
# About the inputs:
import sys, re, os
sys.path.append('..')
from glmgen import milToGridlab
from glmgen import Milsoft_GridLAB_D_Feeder_Generation
from glmgen import calibrateFeeder
from glmgen import processSCADA #TODO: write this module
from glmgen import feeder
from glmgen import AddTapeObjects
#import writeFeederConfig #TODO: future work (after June 1 deadline)
from glmgen import four_node_commercial # feeder dictionary just for testing
from glmgen import four_node_residential # feeder dictionary just for testing

# MilSoft model ( milsoft = [stdPath, seqPath])
# Case Flag
# Feeder Info (region, ratings, etc) TBD
# SCADA Data
# feeder configuration file (user may input feeder configuration file as created by a previous calibration attempt on this feeder)
# feeder calibration file (user may input calibration file as created by a previous calibration attempt on this feeder)
# user_flag_to_calibrate (user may set to 0 to indicate they don't want to calibrate, just want general popualted model)

def createFeederDirectory(feederID):
	# check that this feederID is unique
	# create folder with all the fixin's for this model to be populated and/or calibrated and ran in GLD
	# TODO: directory needs to include subdirectory 'winners' for storing the best .glms from each calibration round
	# TODO: anything with suffix 'glm' might be deleted during calibration since we cleanup unecessary file between rounds. Put schedules, other .glms, etc into a subdirectory if we don't want them deleted. 
	# TODO: If using .csv files for recorder output (as opposed to mysql), add subdirectory 'csv_output' ( or go into makeGLM.py and change the file path -- that's where it's defined. Also would have to change it in gleanMetrics.py where each file is opened.)
	#directory = 'C:\\Users\\d3y051\\Documents\\NRECA_feeder_calibration_2-2013\\Calibration\\github\\omf\\calibration\\Feeder_Test'
	#directory = 'C:\\Users\\d3y051\\Desktop\\calibration\\calibration\\Feeder_Test'
	currentDirectory = os.getcwd()
	workingDirectory = os.path.join(currentDirectory,'workingDir')
	if not os.path.isdir(workingDirectory):
		os.mkdir('workingDir')
	return workingDirectory
	
def OMFmain(milsoft, feeder_info, scada, case_flag, feeder_config, calibration_config, model_name='Feeder', user_flag_to_calibrate=1):

	if milsoft is None:
		print ("Please input a model to convert!")
		# error
		return
	else:
		internal_flag_to_calibrate = 0
		if feeder_config is None:
			if feeder_info is None:
				pass  # We'll use defaults.
			else:
				# use whatever feeder_info is to write a new feeder_config file that will populate Configuration.py appropriatly.
				#feeder_config = writeFeederConfig(feeder_info)  #TODO: This function is in the future. 
				pass
		else:
			pass
			# We should check that this file is a valid feeder_config file for use inside Configuration.py. This is important if it's coming from the user incase they made change to the file that aren't compatible with our format. 
		if calibration_config is None:
			if scada is None:
				pass  # Well, we can't do any calibration but we can still pump out a populated model by using defaults.
			else:
				internal_flag_to_calibrate = 1
				days, SCADA = processSCADA.getValues(scada)
		else:
			pass
			# We'll check that this file is a valid calibration_config file for use inside Configuration.py. This is important if it's coming from the user incase they made change to the file that aren't compatible with our format.
			# This might be where we can find out if this is the final calibration file or if we're starting mid-calibration. To do this maybe:
			#  a. We could have another input to the script that is a flag indicating "yes, this is calibrated" or "no, this isn't finished yet".
			#  b. We could just run the script using this calbiration_config file and let the script determine whether it's calibrated -- this might take too long though. 
			#  c. There could be a flag within the calibration_config file that was set when it was previously determined to be the "best".
		
		# Create base .glm dictionary from milsoft model 
	
		# stdPath, seqPath = milsoft[0], milsoft[1]
		# outGlm = milToGridlab.convert(stdPath,seqPath)
		# directory = createFeederDirectory(milsoft)
		
		# testing with pre-made dictionary
		outGLM = milsoft
		directory = createFeederDirectory(None)
		
		# write base .glm to file (save as .txt so that it isn't run when batch file executed)
		basefile = open(directory+'\\'+model_name+'_base_glm.txt','w')
		basefile.write('\\\\ Base feeder model generated by milToGridlab.py.\n')
		basefile.write(str(outGLM))
		basefile.close()
		
		if internal_flag_to_calibrate == 1 and user_flag_to_calibrate == 1:  # The user must want to calibrate (user_flag_to_calibrate = 1) and we must have SCADA input (internal_flag_to_calibrate = 1).
			# Send base .glm dictionary to calibration function
			final_calib_file, final_dict, last_key = calibrateFeeder.calibrateFeeder(outGLM, days, SCADA, case_flag,feeder_config, calibration_config, directory)
		else:
			# Populate the feeder. 
			print ("Either the user selected not to calibrate this feeder, the SCADA was not input, or this feeder has already been calibrated.")
			final_dict, last_key = Milsoft_GridLAB_D_Feeder_Generation.GLD_Feeder(outGLM,case_flag,directory,calibration_config,feeder_config)
		#AddTapeObjects
		#filename = 'test_feeder'
		if final_dict is not None:
			#AddTapeObjects.add_recorders(final_dict,None,last_key,None,1,0,filename,None,0,0)
			dict_with_recorders, last_key = AddTapeObjects.add_recorders(final_dict,case_flag,0,1,model_name,last_key)
		
			#TODO: Turn final_dict into a .glm and return it to the user. 
			glmstring = str(dict_with_recorders)
			
			if final_calib_file is not None:
				#calib = re.sub('.txt$','',final_calib_file)
				m = re.match(r'^Calib_ID(\d*)_Config_ID(\d*)',final_calib_file)
				if m is not None:
					calib = m.group()
				else:
					calib = 'unrecognized'
			else:
				calib = 'using_default_configuration'
				
			file = open(directory+'\\'+model_name+'_'+calib+'_final.glm', 'w')
			file.write(glmstring)
			file.close()
			print ("Final *.glm file is " + directory + '\\' + model_name + '_' + calib + '_final.glm')
			print ("Calibration file is " + directory + '\\' + final_calib_file)
		else:
			pass
		
def main():
	# test run with pre-made dictionary simuating output from milToGridlab.py. All other parameters are None. Note that SCADA input just needs to be not None for this test-- the values are imput directily into processSCADA.py rather than actually being taken from a file right now.
	glm_object_dict = {}
	glm_object_dict[0] = {'object' : 'overhead_line_conductor',
						  'name' : 'ohlc_100',
						  'geometric_mean_radius' : '0.0244',
						  'resistance' : '0.306'}

	glm_object_dict[1] = {'object' : 'overhead_line_conductor',
						  'name' : 'ohlc_101',
						  'geometric_mean_radius' : '0.00814',
						  'resistance' : '0.592'}

	glm_object_dict[2] = {'object' : 'line_spacing',
						  'name' : 'ls_200',
						  'distance_AB' : '2.5',
						  'distance_BC' : '4.5',
						  'distance_AC' : '7.0',
						  'distance_AN' : '5.656854',
						  'distance_BN' : '4.272002',
						  'distance_CN' : '5.0'}

	glm_object_dict[3] = {'object' : 'line_configuration',
						  'name' : 'lc_300',
						  'conductor_A' : 'ohlc_100',
						  'conductor_B' : 'ohlc_100',
						  'conductor_C' : 'ohlc_100',
						  'conductor_N' : 'ohlc_101',
						  'spacing' : 'ls_200'}

	glm_object_dict[4] = {'object' : 'transformer_configuration',
						  'name' : 'tc_400',
						  'connect_type' : '1',
						  'power_rating' : '7000',
						  'powerA_rating' : '875',
						  'powerB_rating' : '1750',
						  'powerC_rating' : '4375',
						  'primary_voltage' : '7200',
						  'secondary_voltage' : '2400',
						  'resistance' : '0.01',
						  'reactance' : '0.06'}

	glm_object_dict[5] = {'object' : 'node',
						  'name' : 'node1',
						  'phases' : 'ABCN',
                          'bustype' : 'SWING',
						  'voltage_A' : '+7199.558+0.000j',
						  'voltage_B' : '-3599.779-6235.000j',
						  'voltage_C' : '-3599.779+6235.000j',
						  'nominal_voltage' : '7200'}

	glm_object_dict[6] = {'object' : 'overhead_line',
						  'name' : 'oh_12',
						  'phases' : 'ABCN',
						  'from' : 'node1',
						  'to' : 'node2',
						  'length' : '2000',
						  'configuration' : 'lc_300'}

	glm_object_dict[7] = {'object' : 'node',
						  'name' : 'node2',
						  'phases' : 'ABCN',
						  'voltage_A' : '+7199.558+0.000j',
						  'voltage_B' : '-3599.779-6235.000j',
						  'voltage_C' : '-3599.779+6235.000j',
						  'nominal_voltage' : '7200'}

	glm_object_dict[8] = {'object' : 'transformer',
						  'name' : 't_23',
						  'phases' : 'ABCN',
						  'from' : 'node2',
						  'to' : 'node3',
						  'configuration' : 'tc_400'}

	glm_object_dict[9] = {'object' : 'node',
						  'name' : 'node3',
						  'phases' : 'ABCN',
						  'voltage_A' : '+2401.777+0.000j',
						  'voltage_B' : '-1200.889-2080.000j',
						  'voltage_C' : '-1200.889+2080.000j',
						  'nominal_voltage' : '2400'}

	glm_object_dict[10] = {'object' : 'overhead_line',
						  'name' : 'oh_34',
						  'phases' : 'ABCN',
						  'from' : 'node3',
						  'to' : 'node4',
						  'length' : '2500',
						  'configuration' : 'lc_300'}

	glm_object_dict[11] = {'object' : 'node',
						  'name' : 'node4',
						  'phases' : 'ABCN',
						  'voltage_A' : '+2401.777+0.000j',
						  'voltage_B' : '-1200.889-2080.000j',
						  'voltage_C' : '-1200.889+2080.000j',
						  'nominal_voltage' : '2400'}
# Residential load objects:
	glm_object_dict[15] = {'object' : 'transformer',
						'name' : 'SPCT_A_n4-tn4',
						'phases' : 'AS',
						'from' : 'node4',
						'to' : 'tn4A',
						'configuration' : 'SPCT_config_A_500k'}

	glm_object_dict[16] = {'object' : 'transformer',
						'name' : 'SPCT_B_n4-tn4',
						'phases' : 'BS',
						'from' : 'node4',
						'to' : 'tn4B',
						'configuration' : 'SPCT_config_B_500k'}

	glm_object_dict[17] = {'object' : 'transformer',
						'name' : 'SPCT_C_n4-tn4',
						'phases' : 'CS',
						'from' : 'node4',
						'to' : 'tn4C',
						'configuration' : 'SPCT_config_C_667k'}

	glm_object_dict[18] = {'object' : 'triplex_node',
						'name' : 'tn4A',
						'phases' : 'AS',
						'power_12' : '318750.000+197544.508j',
						'nominal_voltage' : '120'}

	glm_object_dict[19] = {'object' : 'triplex_node',
						'name' : 'tn4B',
						'phases' : 'BS',
						'power_12' : '450000.000+217945.947j',
						'nominal_voltage' : '120'}

	glm_object_dict[20] = {'object' : 'triplex_node',
						'name' : 'tn4C',
						'phases' : 'CS',
						'power_12' : '593750.000+195156.187j',
						'nominal_voltage' : '120'}
# Commercial Load Objects:
	# glm_object_dict[12] = {'object' : 'load',
						  # 'parent' : 'node4',
						  # 'name' : 'l4A',
						  # 'phases' : 'AN',
						  # 'constant_power_A' : '785369.750+258138.553j',
						  # 'load_class' : 'C',
						  # 'nominal_voltage' : '2400'}

	# glm_object_dict[13] = {'object' : 'load',
						  # 'parent' : 'node4',
						  # 'name' : 'l4B',
						  # 'phases' : 'BN',
						  # 'constant_power_B' : '1570739.500+516277.107j',
						  # 'load_class' : 'C',
						  # 'nominal_voltage' : '2400'}

	# glm_object_dict[14] = {'object' : 'load',
						  # 'parent' : 'node4',
						  # 'name' : 'l4C',
						  # 'phases' : 'CN',
						  # 'constant_power_C' : '3926848.750+1290692.768j',
						  # 'load_class' : 'C',
						  # 'nominal_voltage' : '2400'}

	#milsoft = four_node_commercial.glm_object_dict
	milsoft = milToGridlab.convert('IEEE 13.std','IEEE 13.seq')
	
	model_name = 'testing_model_IEEE13'
	feeder_info = None # place holder for future feeder information input from the user. 
	scada = 'make_believe_scada_file.xlsx' # place holder for file with scada that will be processed for certain values. 
	case_flag = -1 # base case, do not put None
	feeder_config = None # only used in the case that the user already has a feeder configuration .py file that works (i.e. they've calibrated feeder  or at least part of it)
	calibration_config = None # same as for feeder_config
	
	OMFmain(milsoft, feeder_info, scada, case_flag, feeder_config, calibration_config, model_name, 1) 

if __name__ ==  '__main__':
	main()